customerio/customerio-ios
# chore: after close button pressed show next message in local queue

https://linear.app/customerio/issue/MBL-313/after-close-action-pressed-replace-in-app-message-with-next-in-queue 

Problem to solve: 
As a marketer, I want to display another in-app message inline when close button pressed so I can rotate between multiple in-app messages.

Solution: 
When close button is clicked on an inline View, check the local queue to see if there is another message to display for the element id. 

If there is another message...
1. Remove the current WebView from the inline View and cleanup resources. 
2. Begin rendering the next message. 
3. When next message ready to display, display it by animating the height to match the new message to display. 

If there is not another message, dismiss the inline View. 

Not in scope: 
* When you click the close button and there is another message to display, the currently shown WebView is instantly removed from the UI. It doesn't look the best. Follow-up PRs will add [animations to make this transition look nicer](https://linear.app/customerio/issue/MBL-313/after-close-action-pressed-replace-in-app-message-with-next-in-queue#comment-17c168a1). 

Testing: 
* I wrote automated tests for all use cases I could think of. 
* I tested in UIKit sample app. QA test case: 
1. Send 2 in-app messages to a device with same element id. 
2. You should see 1 message visible on the screen. Click the close button on the message. 
3. You should see the message instantly disappear. ~1-2 seconds later, after the next message is rendered, you should see the next message animate into visiblity on the screen. 
4. Click the close button 1 more time. You should see the inline View dismiss by animating height to 0. 


Here is a demo video showing this off: 



https://github.com/customerio/customerio-ios/assets/2041082/a167dbac-d077-4814-bc01-37f89a9d2c09
